File: Documentation/3.0/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/Querying/HowToPerformDynamicAggregation.cs
using System.Linq;

using Raven.Abstractions.Data;
using Raven.Client;
using Raven.Client.Document;

namespace Raven.Documentation.Samples.ClientApi.Session.Querying
{
	public class HowToPerformDynamicAggregation
	{
		private class Order
		{
			public string CompanyId { get; set; }

			public string CustomerId { get; set; }

			public decimal TotalPrice { get; set; }
		}

		private interface IFoo
		{
			/*
			#region aggregate_1
			DynamicAggregationQuery<TResult> AggregateBy<TResult>(
				this IQueryable<TResult> queryable,
				string path,
				string displayName = null) { ... }

			DynamicAggregationQuery<TResult> AggregateBy<TResult>(
				this IQueryable<TResult> queryable,
				Expression<Func<TResult, object>> path) { ... }

			DynamicAggregationQuery<TResult> AggregateBy<TResult>(
				this IQueryable<TResult> queryable,
				Expression<Func<TResult, object>> path,
				string displayName) { ... }
			#endregion
			*/
		}

		public HowToPerformDynamicAggregation()
		{
			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region aggregate_2
					// sum up all order prices
					// for each customer
					// where price is higher than 500
					FacetResults aggregationResults = session.Query<Order>("Orders/All")
						.Where(x => x.TotalPrice > 500)
						.AggregateBy(x => x.CustomerId)
							.SumOn(x => x.TotalPrice)
						.ToList();

					FacetResult customerAggregation = aggregationResults.Results["CustomerId"];
					double? sum = customerAggregation.Values[0].Sum;
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region aggregate_3
					// count all orders
					// for each customer
					// where price is higher than 500
					FacetResults aggregationResults = session.Query<Order>("Orders/All")
						.Where(x => x.TotalPrice > 500)
						.AggregateBy(x => x.CustomerId)
							.CountOn(x => x.TotalPrice)
						.ToList();

					FacetResult customerAggregation = aggregationResults.Results["CustomerId"];
					int? count = customerAggregation.Values[0].Count;
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region aggregate_4
					// count price average for orders
					// for each customer
					// where price is higher than 500
					FacetResults aggregationResults = session.Query<Order>("Orders/All")
						.Where(x => x.TotalPrice > 500)
						.AggregateBy(x => x.CustomerId)
							.AverageOn(x => x.TotalPrice)
						.ToList();

					FacetResult customerAggregation = aggregationResults.Results["CustomerId"];
					double? average = customerAggregation.Values[0].Average;
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region aggregate_5
					// count max and min price for orders
					// for each customer
					// where price is higher than 500
					FacetResults aggregationResults = session.Query<Order>("Orders/All")
						.Where(x => x.TotalPrice > 500)
						.AggregateBy(x => x.CustomerId)
							.MinOn(x => x.TotalPrice)
							.MaxOn(x => x.TotalPrice)
						.ToList();

					FacetResult customerAggregation = aggregationResults.Results["CustomerId"];
					double? min = customerAggregation.Values[0].Min;
					double? max = customerAggregation.Values[0].Max;
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region aggregate_6
					FacetResults aggregationResults = session.Query<Order>("Orders/All")
						.AggregateBy(x => x.CustomerId)
							.AddRanges(x => x.TotalPrice < 100)
							.AddRanges(x => x.TotalPrice >= 100 && x.TotalPrice < 500)
							.AddRanges(x => x.TotalPrice >= 500 && x.TotalPrice < 1000)
							.AddRanges(x => x.TotalPrice >= 1000)
						.ToList();

					FacetResult customerAggregation = aggregationResults.Results["CustomerId"];
					FacetValue range1 = customerAggregation.Values.First(x => x.Range == "[NULL to Dx100]");
					FacetValue range2 = customerAggregation.Values.First(x => x.Range == "{Dx100 to Dx500]");
					FacetValue range3 = customerAggregation.Values.First(x => x.Range == "{Dx500 to Dx1000]");
					FacetValue range4 = customerAggregation.Values.First(x => x.Range == "{Dx1000 to NULL]");
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region aggregate_7
					// sum up all order prices
					// for each customer
					// count price average for all orders
					// for each company
					FacetResults aggregationResults = session.Query<Order>("Orders/All")
						.AggregateBy(x => x.CustomerId)
							.SumOn(x => x.TotalPrice)
						.AndAggregateOn(x => x.CompanyId)
							.AverageOn(x => x.TotalPrice)
						.ToList();

					FacetResult customerAggregation = aggregationResults.Results["CustomerId"];
					FacetResult companyAggregation = aggregationResults.Results["CompanyId"];

					double? sum = customerAggregation.Values[0].Sum;
					double? average = companyAggregation.Values[0].Average;
					#endregion
				}
			}
		}
	}
}

File: Documentation/3.5/Samples/java/src/test/java/net/ravendb/clientapi/session/querying/HowToPerformDynamicAggregation.java
package net.ravendb.clientapi.session.querying;

import java.util.List;

import com.mysema.query.annotations.QueryEntity;
import com.mysema.query.types.Path;

import net.ravendb.abstractions.data.FacetResult;
import net.ravendb.abstractions.data.FacetResults;
import net.ravendb.abstractions.data.FacetValue;
import net.ravendb.client.IDocumentSession;
import net.ravendb.client.IDocumentStore;
import net.ravendb.client.document.DocumentStore;
import net.ravendb.client.linq.DynamicAggregationQuery;


public class HowToPerformDynamicAggregation {
  @QueryEntity
  public static class Order {
    private String companyId;
    private String customerId;
    private double totalPrice;

    public String getCompanyId() {
      return companyId;
    }
    public void setCompanyId(String companyId) {
      this.companyId = companyId;
    }
    public String getCustomerId() {
      return customerId;
    }
    public void setCustomerId(String customerId) {
      this.customerId = customerId;
    }
    public double getTotalPrice() {
      return totalPrice;
    }
    public void setTotalPrice(double totalPrice) {
      this.totalPrice = totalPrice;
    }
  }

  @SuppressWarnings("unused")
  private interface IFoo<T> {
    //region aggregate_1
    public DynamicAggregationQuery<T> aggregateBy(String path);

    public DynamicAggregationQuery<T> aggregateBy(String path, String displayName);

    public DynamicAggregationQuery<T> aggregateBy(Path<?> path);

    public DynamicAggregationQuery<T> aggregateBy(Path<?> path, String displayName);
    //endregion
  }

  @SuppressWarnings({"unused", "boxing"})
  public HowToPerformDynamicAggregation() throws Exception {
    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region aggregate_2
        // sum up all order prices
        // for each customer
        // where price is higher than 500
        QHowToPerformDynamicAggregation_Order o = QHowToPerformDynamicAggregation_Order.order;
        FacetResults aggregationResults = session
          .query(Order.class, "Orders/All")
          .where(o.totalPrice.gt(500))
          .aggregateBy(o.customerId)
            .sumOn(o.totalPrice)
          .toList();

        FacetResult customerAggregation = aggregationResults.getResults().get("CustomerId");
        Double sum = customerAggregation.getValues().get(0).getSum();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region aggregate_3
        // count all orders
        // for each customer
        // where price is higher than 500
        QHowToPerformDynamicAggregation_Order o = QHowToPerformDynamicAggregation_Order.order;
        FacetResults aggregationResults = session
          .query(Order.class, "Orders/All")
          .where(o.totalPrice.gt(500))
          .aggregateBy(o.customerId)
            .countOn(o.totalPrice)
          .toList();

        FacetResult customerAggregation = aggregationResults.getResults().get("CustomerId");
        Integer count = customerAggregation.getValues().get(0).getCount();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region aggregate_4
        // count price average for orders
        // for each customer
        // where price is higher than 500
        QHowToPerformDynamicAggregation_Order o = QHowToPerformDynamicAggregation_Order.order;
        FacetResults aggregationResults = session
          .query(Order.class, "Orders/All")
          .where(o.totalPrice.gt(500))
          .aggregateBy(o.customerId)
            .averageOn(o.totalPrice)
          .toList();

        FacetResult customerAggregation = aggregationResults.getResults().get("CustomerId");
        Double average = customerAggregation.getValues().get(0).getAverage();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region aggregate_5
        // count max and min price for orders
        // for each customer
        // where price is higher than 500
        QHowToPerformDynamicAggregation_Order o = QHowToPerformDynamicAggregation_Order.order;
        FacetResults aggregationResults = session
          .query(Order.class, "Orders/All")
          .where(o.totalPrice.gt(500))
          .aggregateBy(o.customerId)
            .minOn(o.totalPrice)
            .maxOn(o.totalPrice)
          .toList();

        FacetResult customerAggregation = aggregationResults.getResults().get("CustomerId");
        Double min = customerAggregation.getValues().get(0).getMin();
        Double max = customerAggregation.getValues().get(0).getMax();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region aggregate_6
        QHowToPerformDynamicAggregation_Order o = QHowToPerformDynamicAggregation_Order.order;
        FacetResults aggregationResults = session.query(Order.class, "Orders/All")
          .aggregateBy(o.customerId)
            .addRanges(o.totalPrice.lt(10))
            .addRanges(o.totalPrice.goe(100).and(o.totalPrice.lt(500)))
            .addRanges(o.totalPrice.goe(500).and(o.totalPrice.lt(1000)))
            .addRanges(o.totalPrice.goe(1000))
          .toList();

        FacetResult customerAggregation = aggregationResults.getResults().get("CustomerId");
        List<FacetValue> values = customerAggregation.getValues();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region aggregate_7
        // sum up all order prices
        // for each customer
        // count price average for all orders
        // for each company
        QHowToPerformDynamicAggregation_Order o = QHowToPerformDynamicAggregation_Order.order;
        FacetResults aggregationResults = session
          .query(Order.class, "Orders/All")
          .aggregateBy(o.customerId)
            .sumOn(o.totalPrice)
          .andAggregateOn(o.companyId)
            .averageOn(o.totalPrice)
          .toList();

        FacetResult customerAggregation = aggregationResults.getResults().get("CustomerId");
        FacetResult companyAggregation = aggregationResults.getResults().get("CompanyId");

        Double sum = customerAggregation.getValues().get(0).getSum();
        Double average = companyAggregation.getValues().get(0).getAverage();
        //endregion
      }
    }
  }
}
